core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,546 @@
+        getResource(o.getResourcesList, "mem"))
+    }
+    logTrace(s"Received offers from Mesos: \n${printOffers(currentOffers)}")
+    val tasks = new mutable.HashMap[OfferID, ArrayBuffer[TaskInfo]]()
+    val currentTime = new Date()
+    def scheduleTasks(
+        tasksFunc: () => Seq[(MesosDriverDescription, Option[RetryState])],
+        scheduledCallback: (String) => Unit): Unit = {
+      val candidates = tasksFunc()
+      for ((submission, retryState) <- candidates) {
+        val driverCpu = submission.cores
+        val driverMem = submission.mem
+        logTrace(s"Finding offer to launch driver with cpu: $driverCpu, mem: $driverMem")
+        val offerOption = currentOffers.find { o =>
+          o.cpu >= driverCpu && o.mem >= driverMem
+        }
+
+        if (offerOption.isEmpty) {
+          logDebug(s"Unable to find offer to launch driver id: ${submission.submissionId.get}," +
+              s"cpu: $driverCpu, mem: $driverMem")
+        } else {
+          val offer = offerOption.get
+          offer.cpu -= driverCpu
+          offer.mem -= driverMem
+          val taskId = TaskID.newBuilder().setValue(submission.submissionId.get).build()
+          val cpuResource = Resource.newBuilder()
+            .setName("cpus").setType(Value.Type.SCALAR)
+            .setScalar(Value.Scalar.newBuilder().setValue(driverCpu)).build()
+          val memResource = Resource.newBuilder()
+            .setName("mem").setType(Value.Type.SCALAR)
+            .setScalar(Value.Scalar.newBuilder().setValue(driverMem)).build()
+          val commandInfo = buildCommand(submission)
+          val appName = submission.schedulerProperties("spark.app.name")
+          val taskInfo = TaskInfo.newBuilder()
+            .setTaskId(taskId)
+            .setName(s"Driver for $appName")
+            .setSlaveId(offer.offer.getSlaveId)
+            .setCommand(commandInfo)
+            .addResources(cpuResource)
+            .addResources(memResource)
+            .build
+          val queuedTasks = if (!tasks.contains(offer.offer.getId)) {
+            val buffer = new ArrayBuffer[TaskInfo]
+            tasks(offer.offer.getId) = buffer
+            buffer
+          } else {
+            tasks(offer.offer.getId)
+          }
+          queuedTasks += taskInfo
+          logTrace(s"Using offer ${offer.offer.getId.getValue} to launch driver " +
+            submission.submissionId.get)
+
+          launchedDrivers.set(
+            submission.submissionId.get,
+            new MesosClusterTaskState(submission, taskId, offer.offer.getSlaveId,
+              None, new Date(), retryState))
+          scheduledCallback(submission.submissionId.get)
+        }
+      }
+    }
+
+    stateLock.synchronized {
+      scheduleTasks(() => {
+        superviseRetryList.getNextRetries(currentTime)
+      }, (id: String) => {
+        superviseRetryList.remove(id)
+      })
+      scheduleTasks(() => queue.drivers.map(d => (d, None)), (id) => queue.remove(id))
+    }
```

Comment:
style: this whole block is hard to read because of all the curly braces and the `()`. I think even just rewriting it this way will make it much easier to follow.
```
stateLock.synchronized {
  scheduleTasks(
    () => superviseRetryList.getNextRetries(currentTime),
    (id: String) => superviseRetryList.remove(id))
  scheduleTasks(
    () => queue.drivers.map(d => (d, None)),
    (id) => queue.remove(id))
}
```